### ISSUE TITLE
Enabled sticky sessions flag not effective when load-balancing method is missing or invalid.

### ISSUE THREAD

**Issue by kleinsasserm** [containous/traefik#1568]
Hello!

The documentation on stickiness on https://docs.traefik.io/basics/ says:
```
Sticky sessions are supported with both load balancers. When sticky sessions are enabled, a cookie called _TRAEFIK_BACKEND is set on the initial request. On subsequent requests, the client will be directed to the backend stored in the cookie if it is still healthy. If not, a new backend will be assigned.
```
and I guess this is not correct anymore. If I leave the default (wrr) I don't receive a loadbalancing cookie in the browser when setting sticky = true. It is only working when method = "drr".

Also, but maybe I'am wrong, I thought that there was an indicator on the web ui which was indicating that stickiness is enabled?

![traefik-web-sticky](https://cloud.githubusercontent.com/assets/9364572/25836181/1223caa8-3485-11e7-8aca-42d4c28629eb.jpg)

### Output of `traefik version`: (_What version of Traefik are you using?_)
Version:      v1.2.3
Codename:     morbier
Go version:   go1.7.5
Built:        2017-04-13_07:32:59PM
OS/Arch:      linux/amd64

**Comment by timoreimann** [comment]
Servus Mario,

just had a peek into the code: AFAICS, sticky sessions _should_ work with both strategies.

When you start Traefik with DEBUG level logging enabled, do you see a message like 

"Sticky sessions with cookie <cookie name>"

? This should come up.

**Comment by kleinsasserm** [comment]
Servus Timo,

it does not work if I only specify the sticky = true option:

__not working__:

```
DEBU[2017-05-09T07:54:30+02:00] Load balancer method '&{Method: Sticky:true}' for backend backend1: Invalid method, using default. Using default wrr.
```

```
INFO[2017-05-09T07:54:30+02:00] Traefik version v1.2.3 built on 2017-04-13_07:32:59PM 
INFO[2017-05-09T07:54:30+02:00] Using TOML configuration file /tmp/conf.toml 
DEBU[2017-05-09T07:54:30+02:00] Global configuration loaded {"GraceTimeOut":10,"Debug":true,"CheckNewVersion":true,"AccessLogsFile":"","TraefikLogsFile":"","LogLevel":"DEBUG","EntryPoints":{"http":{"Network":"","Address":":9192","TLS":null,"Redirect":null,"Auth":null,"Compress":false}},"Cluster":null,"Constraints":[],"ACME":null,"DefaultEntryPoints":[],"ProvidersThrottleDuration":2000000000,"MaxIdleConnsPerHost":200,"InsecureSkipVerify":false,"Retry":null,"Docker":null,"File":{"Watch":true,"Filename":"/tmp/conf.toml","Constraints":null},"Web":{"Address":":9191","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Auth":null},"Marathon":null,"Consul":null,"ConsulCatalog":null,"Etcd":null,"Zookeeper":null,"Boltdb":null,"Kubernetes":null,"Mesos":null,"Eureka":null,"ECS":null,"Rancher":null} 
INFO[2017-05-09T07:54:30+02:00] Preparing server http &{Network: Address::9192 TLS:<nil> Redirect:<nil> Auth:<nil> Compress:false} 
INFO[2017-05-09T07:54:30+02:00] Starting provider *provider.File {"Watch":true,"Filename":"/tmp/conf.toml","Constraints":null} 
INFO[2017-05-09T07:54:30+02:00] Starting provider *main.WebProvider {"Address":":9191","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Auth":null} 
INFO[2017-05-09T07:54:30+02:00] Starting server on :9192                     
DEBU[2017-05-09T07:54:30+02:00] Load balancer method '&{Method: Sticky:true}' for backend backend1: Invalid method, using default. Using default wrr. 
DEBU[2017-05-09T07:54:30+02:00] Configuration received from provider file: {"backends":{"backend1":{"servers":{"server0":{"url":"http://127.0.0.1:4444","weight":1},"server1":{"url":"http://127.0.0.1:4445","weight":1}},"loadBalancer":{"method":"wrr"}}},"frontends":{"frontend1":{"entryPoints":["http"],"backend":"backend1","priority":0}}} 
DEBU[2017-05-09T07:54:30+02:00] Last file config received more than 2s, OK   
DEBU[2017-05-09T07:54:30+02:00] Creating frontend frontend1                  
DEBU[2017-05-09T07:54:30+02:00] Wiring frontend frontend1 to entryPoint http 
DEBU[2017-05-09T07:54:30+02:00] Creating backend backend1                    
DEBU[2017-05-09T07:54:30+02:00] Creating load-balancer wrr                   
DEBU[2017-05-09T07:54:30+02:00] Creating server server0 at http://127.0.0.1:4444 with weight 1 
DEBU[2017-05-09T07:54:30+02:00] Creating server server1 at http://127.0.0.1:4445 with weight 1 
INFO[2017-05-09T07:54:30+02:00] Server configuration reloaded on :9192  
```
It **does** work when I specify method = "drr" **and** it works when I manually set method = "wrr". Seems like that a default condition is missing.

__working__:

```
INFO[2017-05-09T08:39:09+02:00] Traefik version v1.2.3 built on 2017-04-13_07:32:59PM 
INFO[2017-05-09T08:39:09+02:00] Using TOML configuration file /tmp/conf.toml 
DEBU[2017-05-09T08:39:09+02:00] Global configuration loaded {"GraceTimeOut":10,"Debug":true,"CheckNewVersion":true,"AccessLogsFile":"","TraefikLogsFile":"","LogLevel":"DEBUG","EntryPoints":{"http":{"Network":"","Address":":9192","TLS":null,"Redirect":null,"Auth":null,"Compress":false}},"Cluster":null,"Constraints":[],"ACME":null,"DefaultEntryPoints":[],"ProvidersThrottleDuration":2000000000,"MaxIdleConnsPerHost":200,"InsecureSkipVerify":false,"Retry":null,"Docker":null,"File":{"Watch":true,"Filename":"/tmp/conf.toml","Constraints":null},"Web":{"Address":":9191","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Auth":null},"Marathon":null,"Consul":null,"ConsulCatalog":null,"Etcd":null,"Zookeeper":null,"Boltdb":null,"Kubernetes":null,"Mesos":null,"Eureka":null,"ECS":null,"Rancher":null} 
INFO[2017-05-09T08:39:09+02:00] Preparing server http &{Network: Address::9192 TLS:<nil> Redirect:<nil> Auth:<nil> Compress:false} 
INFO[2017-05-09T08:39:09+02:00] Starting provider *provider.File {"Watch":true,"Filename":"/tmp/conf.toml","Constraints":null} 
INFO[2017-05-09T08:39:09+02:00] Starting provider *main.WebProvider {"Address":":9191","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Auth":null} 
INFO[2017-05-09T08:39:09+02:00] Starting server on :9192                     
DEBU[2017-05-09T08:39:09+02:00] Configuration received from provider file: {"backends":{"backend1":{"servers":{"server0":{"url":"http://127.0.0.1:4444","weight":1},"server1":{"url":"http://127.0.0.1:4445","weight":1}},"loadBalancer":{"method":"drr","sticky":true}}},"frontends":{"frontend1":{"entryPoints":["http"],"backend":"backend1","priority":0}}} 
DEBU[2017-05-09T08:39:09+02:00] Last file config received more than 2s, OK   
DEBU[2017-05-09T08:39:09+02:00] Creating frontend frontend1                  
DEBU[2017-05-09T08:39:09+02:00] Wiring frontend frontend1 to entryPoint http 
DEBU[2017-05-09T08:39:09+02:00] Creating backend backend1                    
DEBU[2017-05-09T08:39:09+02:00] Creating load-balancer drr                   
DEBU[2017-05-09T08:39:09+02:00] Sticky session with cookie _TRAEFIK_BACKEND  
DEBU[2017-05-09T08:39:09+02:00] Creating server server0 at http://127.0.0.1:4444 with weight 1 
DEBU[2017-05-09T08:39:09+02:00] Creating server server1 at http://127.0.0.1:4445 with weight 1 
INFO[2017-05-09T08:39:09+02:00] Server configuration reloaded on :9192   

   
INFO[2017-05-09T08:43:06+02:00] Traefik version v1.2.3 built on 2017-04-13_07:32:59PM 
INFO[2017-05-09T08:43:06+02:00] Using TOML configuration file /tmp/conf.toml 
DEBU[2017-05-09T08:43:06+02:00] Global configuration loaded {"GraceTimeOut":10,"Debug":true,"CheckNewVersion":true,"AccessLogsFile":"","TraefikLogsFile":"","LogLevel":"DEBUG","EntryPoints":{"http":{"Network":"","Address":":9192","TLS":null,"Redirect":null,"Auth":null,"Compress":false}},"Cluster":null,"Constraints":[],"ACME":null,"DefaultEntryPoints":[],"ProvidersThrottleDuration":2000000000,"MaxIdleConnsPerHost":200,"InsecureSkipVerify":false,"Retry":null,"Docker":null,"File":{"Watch":true,"Filename":"/tmp/conf.toml","Constraints":null},"Web":{"Address":":9191","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Auth":null},"Marathon":null,"Consul":null,"ConsulCatalog":null,"Etcd":null,"Zookeeper":null,"Boltdb":null,"Kubernetes":null,"Mesos":null,"Eureka":null,"ECS":null,"Rancher":null} 
INFO[2017-05-09T08:43:06+02:00] Preparing server http &{Network: Address::9192 TLS:<nil> Redirect:<nil> Auth:<nil> Compress:false} 
INFO[2017-05-09T08:43:06+02:00] Starting provider *provider.File {"Watch":true,"Filename":"/tmp/conf.toml","Constraints":null} 
INFO[2017-05-09T08:43:06+02:00] Starting provider *main.WebProvider {"Address":":9191","CertFile":"","KeyFile":"","ReadOnly":false,"Statistics":null,"Metrics":null,"Auth":null} 
INFO[2017-05-09T08:43:06+02:00] Starting server on :9192                     
DEBU[2017-05-09T08:43:06+02:00] Configuration received from provider file: {"backends":{"backend1":{"servers":{"server0":{"url":"http://127.0.0.1:4444","weight":1},"server1":{"url":"http://127.0.0.1:4445","weight":1}},"loadBalancer":{"method":"wrr","sticky":true}}},"frontends":{"frontend1":{"entryPoints":["http"],"backend":"backend1","priority":0}}} 
DEBU[2017-05-09T08:43:06+02:00] Last file config received more than 2s, OK   
DEBU[2017-05-09T08:43:06+02:00] Creating frontend frontend1                  
DEBU[2017-05-09T08:43:06+02:00] Wiring frontend frontend1 to entryPoint http 
DEBU[2017-05-09T08:43:06+02:00] Creating backend backend1                    
DEBU[2017-05-09T08:43:06+02:00] Creating load-balancer wrr                   
DEBU[2017-05-09T08:43:06+02:00] Sticky session with cookie _TRAEFIK_BACKEND  
DEBU[2017-05-09T08:43:06+02:00] Creating server server1 at http://127.0.0.1:4445 with weight 1 
DEBU[2017-05-09T08:43:06+02:00] Creating server server0 at http://127.0.0.1:4444 with weight 1 
INFO[2017-05-09T08:43:06+02:00] Server configuration reloaded on :9192
```

**Comment by timoreimann** [comment]
This is actually a bug, and it's already visible in the logs -- in particular, this line:

 `DEBU[2017-05-09T07:54:30+02:00] Load balancer method '&{Method: Sticky:true}' for backend backend1: Invalid method, using default. Using default wrr.`

Traefik thinks the load-balancing method is `Sticky: true`, throws it away and defaults to `wrr`. Only if we specify the method explicitly we get the proper configuration.

It looks like the unmarshaling/decoding/config-parsing part is doing something wrong. Will investigate.

**Comment by timoreimann** [comment]
Ah my bad, I misinterpreted the log line: The Method is actually empty and Sticky is set.

The quest continues...

**Comment by timoreimann** [comment]
Found the problem. Fix to follow.

**Comment by timoreimann** [comment]
Filed fix for 1.3 in PR 1585.

**Comment by kleinsasserm** [comment]
Thank you!

**Comment by ldez** [comment]
Closed by #1585
